proof/refine/ARM/Ipc_R.thy

```diff
@@ -4604,68 +4493,446 @@ lemma maybe_return_sc_weak_valid_sched_action:
-           apply (rule corres_if, simp)
-            apply (rule corres_trivial, simp)
-           apply simp
-          apply (rule corres_trivial, simp)
-         apply (rule threadget_corres, simp add: tcb_relation_def)
-        apply (wp thread_get_wp)
-       apply (wp threadGet_wp)
-      apply (rule threadget_corres, simp add: tcb_relation_def)
-     apply (wp thread_get_wp)
-    apply (wpsimp wp: threadGet_wp)
-   apply (fastforce simp: projectKO_eq projectKO_tcb obj_at'_def)+
+     apply (wpsimp wp: get_simple_ko_wp)
+    apply (wpsimp wp: getReply_wp)
+   apply (clarsimp simp: valid_cap_def)
+   apply (frule (1) valid_objs_ko_at[OF invs_valid_objs])
+   apply (clarsimp simp: valid_obj_def valid_reply_def)
+  apply (clarsimp simp: valid_cap_def)
+  apply (erule cross_relF[OF _ reply_at'_cross_rel])
+  apply fastforce
   done
 
-lemma tcb_ep_dequeue_corres:
-  "qs = qs' \<Longrightarrow> corres (=) \<top> \<top> (tcb_ep_dequeue t qs) (tcbEPDequeue t qs')"
-  by (clarsimp simp: tcb_ep_dequeue_def tcbEPDequeue_def)
+lemma ri_preamble_vbreply:
+  "receive_ipc_preamble_rv p pp s \<Longrightarrow> valid_bound_reply pp s"
+  by (case_tac pp; clarsimp elim!: reply_at_ppred_reply_at)
+
+lemma ri_preamble_not_in_sc:
+  "\<lbrakk>sym_refs (state_refs_of s);
+    valid_replies s;
+    receive_ipc_preamble_rv reply_cap replyOpt s\<rbrakk>
+   \<Longrightarrow> valid_bound_obj (\<lambda>a b. a \<notin> fst ` replies_with_sc b) replyOpt s"
+  apply (case_tac replyOpt; simp)
+  apply (erule (1) no_tcb_not_in_replies_with_sc, simp)
+  done
+
+lemma receiveIPC_corres_helper:
+  "(do replyOpt <-
+                 case replyCap of capability.NullCap \<Rightarrow> return Nothing
+                 | capability.ReplyCap r v18 \<Rightarrow> return (Just r)
+                 | _ \<Rightarrow> haskell_fail [];
+                 y <-
+                 when (\<exists>y. replyOpt = Some y)
+                  (do tptrOpt <- liftM replyTCB (getReply (the replyOpt));
+                      when ((\<exists>y. tptrOpt = Some y) \<and> tptrOpt \<noteq> Some thread)
+                       (cancelIPC (the tptrOpt))
+                   od);
+                 f replyOpt
+          od) = (do replyOpt <-
+                 case replyCap of capability.NullCap \<Rightarrow> return Nothing
+                 | capability.ReplyCap r _ \<Rightarrow>
+                 (do tptrOpt <- liftM replyTCB (getReply (r));
+                     when (tptrOpt \<noteq> Nothing \<and> tptrOpt \<noteq> Some thread) $ cancelIPC (the tptrOpt);
+                     return (Just r)
+                  od)
+                 | _ \<Rightarrow> haskell_fail [];
+                 f replyOpt
+          od)"
+  by (case_tac replyCap; simp add: bind_assoc)
+
+lemma receiveIPC_corres:
+  assumes "is_ep_cap cap" and "cap_relation cap cap'" and "cap_relation reply_cap replyCap"
+      and "is_reply_cap reply_cap \<or> (reply_cap = cap.NullCap)"
+  shows
+    "corres dc (einvs and valid_cap cap
```

Comment:
`einvs` contains `valid_sched`, so you can take that out (unless it's good for the pattern matching, in which case maybe just write `invs`?)